File: backend/app/schemas/cosmetic.py
"""
Cosmetic schemas for cosmetics and inventory management.
Requirements: 3.1, 3.2, 3.3
"""

from datetime import datetime
from typing import Optional, List
from enum import Enum

from pydantic import Field

from app.schemas.base import BaseSchema, TimestampMixin


# ============================================
# Enums
# ============================================

class CosmeticType(str, Enum):
    """Cosmetic item types. Requirements: 3.1"""
    SKIN = "skin"
    EMOTE = "emote"
    BANNER = "banner"
    NAMEPLATE = "nameplate"
    EFFECT = "effect"
    TRAIL = "trail"


class Rarity(str, Enum):
    """Cosmetic rarity tiers. Requirements: 3.2"""
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    EPIC = "epic"
    LEGENDARY = "legendary"


# ============================================
# Cosmetic Schemas
# ============================================

class Cosmetic(BaseSchema):
    """Cosmetic item definition from catalog."""
    
    id: str = Field(..., description="Cosmetic UUID")
    name: str = Field(..., description="Cosmetic name")
    type: CosmeticType = Field(..., description="Cosmetic type")
    rarity: Rarity = Field(..., description="Rarity tier")
    description: Optional[str] = Field(None, description="Item description")
    image_url: str = Field(..., description="Image URL for display")
    preview_video_url: Optional[str] = Field(None, description="Preview video URL")
    price_coins: int = Field(default=0, description="Price in coins")
    price_premium: Optional[int] = Field(None, description="Price in premium currency")
    release_date: datetime = Field(default_factory=datetime.utcnow, description="Release date")
    event: Optional[str] = Field(None, description="Associated event name")
    is_limited: bool = Field(default=False, description="Limited edition flag")
    owned_by_count: int = Field(default=0, description="Number of players who own this")


class CosmeticCreate(BaseSchema):
    """Schema for creating a new cosmetic (admin only)."""
    
    name: str = Field(..., min_length=1, max_length=100)
    type: CosmeticType
    rarity: Rarity
    description: Optional[str] = Field(None, max_length=500)
    image_url: str
    preview_video_url: Optional[str] = None
    price_coins: int = Field(default=0, ge=0)
    price_premium: Optional[int] = Field(None, ge=0)
    event: Optional[str] = Field(None, max_length=50)
    is_limited: bool = False


# ============================================
# Inventory Schemas
# ============================================

class InventoryItem(BaseSchema):
    """Item in player inventory."""
    
    id: str = Field(..., description="Inventory entry UUID")
    cosmetic: Cosmetic = Field(..., description="The cosmetic item")
    acquired_date: datetime = Field(..., description="When the item was acquired")
    is_equipped: bool = Field(default=False, description="Whether item is currently equipped")


class InventoryItemSimple(BaseSchema):
    """Simplified inventory item (without full cosmetic details)."""
    
    id: str
    cosmetic_id: str
    acquired_date: datetime
    is_equipped: bool = False


# ============================================
# Loadout Schemas
# ============================================

class Loadout(BaseSchema):
    """Currently equipped cosmetics."""
    
    user_id: str = Field(..., description="User UUID")
    skin_equipped: Optional[Cosmetic] = Field(None, description="Equipped skin")
    emote_equipped: Optional[Cosmetic] = Field(None, description="Equipped emote")
    banner_equipped: Optional[Cosmetic] = Field(None, description="Equipped banner")
    nameplate_equipped: Optional[Cosmetic] = Field(None, description="Equipped nameplate")
    effect_equipped: Optional[Cosmetic] = Field(None, description="Equipped effect")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")


class LoadoutSimple(BaseSchema):
    """Simplified loadout with just IDs."""
    
    user_id: str
    skin_equipped: Optional[str] = None
    emote_equipped: Optional[str] = None
    banner_equipped: Optional[str] = None
    nameplate_equipped: Optional[str] = None
    effect_equipped: Optional[str] = None


# ============================================
# Shop Schemas
# ============================================

class ShopFilters(BaseSchema):
    """Filters for shop browsing."""
    
    type: Optional[CosmeticType] = Field(None, description="Filter by cosmetic type")
    rarity: Optional[Rarity] = Field(None, description="Filter by rarity")
    event: Optional[str] = Field(None, description="Filter by event")
    is_limited: Optional[bool] = Field(None, description="Filter limited items only")
    max_price: Optional[int] = Field(None, ge=0, description="Maximum price in coins")
    min_price: Optional[int] = Field(None, ge=0, description="Minimum price in coins")


class ShopResponse(BaseSchema):
    """Shop listing response."""
    
    items: List[Cosmetic] = Field(default_factory=list)
    total: int = Field(default=0)
    page: int = Field(default=1)
    page_size: int = Field(default=20)


# ============================================
# Purchase Schemas
# ============================================

class PurchaseRequest(BaseSchema):
    """Request to purchase a cosmetic."""
    
    cosmetic_id: str = Field(..., description="ID of cosmetic to purchase")
    currency: str = Field(default="coins", description="Currency to use: coins or premium")


class PurchaseResponse(BaseSchema):
    """Response after successful purchase."""
    
    success: bool = True
    inventory_item: InventoryItem
    remaining_balance: int = Field(..., description="Remaining currency balance")


# ============================================
# Equip Schemas
# ============================================

class EquipRequest(BaseSchema):
    """Request to equip a cosmetic."""
    
    cosmetic_id: str = Field(..., description="ID of cosmetic to equip")


class UnequipRequest(BaseSchema):
    """Request to unequip a cosmetic slot."""
    
    slot: CosmeticType = Field(..., description="Slot to unequip")


# ============================================
# Inventory Response
# ============================================

class InventoryResponse(BaseSchema):
    """Full inventory response."""
    
    items: List[InventoryItem] = Field(default_factory=list)
    total: int = Field(default=0)
    loadout: Optional[Loadout] = None

File: frontend/src/types/cosmetic.ts
/**
 * Cosmetic types for shop and inventory.
 * Requirements: 3.1, 3.2
 */

/**
 * Types of cosmetic items.
 */
export type CosmeticType = 'skin' | 'emote' | 'banner' | 'nameplate' | 'effect' | 'trail' | 'playercard' | 'runner';

/**
 * Rarity levels for cosmetics.
 */
export type Rarity = 'common' | 'uncommon' | 'rare' | 'epic' | 'legendary';

/**
 * Rarity colors for display.
 */
export const RARITY_COLORS: Record<Rarity, string> = {
  common: '#9ca3af',     // gray
  uncommon: '#22c55e',   // green
  rare: '#3b82f6',       // blue
  epic: '#a855f7',       // purple
  legendary: '#f59e0b',  // orange
};

/**
 * Rarity background gradients.
 */
export const RARITY_GRADIENTS: Record<Rarity, string> = {
  common: 'from-gray-600 to-gray-700',
  uncommon: 'from-green-600 to-green-700',
  rare: 'from-blue-600 to-blue-700',
  epic: 'from-[#a855f7] to-[#9333ea]',
  legendary: 'from-orange-500 to-yellow-600',
};

/**
 * Cosmetic item from catalog.
 */
export interface Cosmetic {
  id: string;
  name: string;
  description?: string;
  type: CosmeticType;
  rarity: Rarity;
  price_coins: number;
  image_url: string;
  shop_preview_url?: string;  // Custom shop preview image (takes priority over image_url)
  preview_video_url?: string;
  model_url?: string;  // 3D GLB model URL for interactive preview
  is_limited: boolean;
  event?: string;
  release_date?: string;
  owned_by_count?: number;
  available_until?: string;
  skin_id?: string;  // Maps to frontend SkinId for sprite-based skins
  // Dynamic asset URLs (from CMS)
  sprite_sheet_url?: string;
  sprite_meta_url?: string;
  is_featured?: boolean;
  sort_order?: number;
}

/**
 * Inventory item (owned cosmetic).
 */
export interface InventoryItem {
  id: string;
  cosmetic_id: string;
  cosmetic: Cosmetic;
  acquired_date: string;
  is_equipped: boolean;
}

/**
 * Player loadout (equipped items).
 */
export interface Loadout {
  skin?: string;
  emote?: string;
  banner?: string;
  nameplate?: string;
  effect?: string;
  trail?: string;
  playercard?: string;
  runner?: string;
}

/**
 * Shop filters.
 */
export interface ShopFilters {
  type?: CosmeticType;
  rarity?: Rarity;
  max_price?: number;
  search?: string;
}

/**
 * Get display name for cosmetic type.
 */
export function getCosmeticTypeName(type: CosmeticType): string {
  const names: Record<CosmeticType, string> = {
    skin: 'Skin',
    emote: 'Emote',
    banner: 'Banner',
    nameplate: 'Nameplate',
    effect: 'Effect',
    trail: 'Trail',
    playercard: 'Player Card',
    runner: 'Runner',
  };
  return names[type];
}

/**
 * Get slot name for loadout.
 */
export function getSlotForType(type: CosmeticType): keyof Loadout {
  return type;
}
